Content.Server/GameObjects/Components/Storage/WallStorageComponent.cs

```diff
@@ -0,0 +1,185 @@
+﻿using System;
+using Content.Server.GameObjects.Components.Weapon.Ranged.Barrels;
+using Content.Server.GameObjects.EntitySystems;
+using Content.Shared.GameObjects;
+using Content.Shared.GameObjects.Components.Storage;
+using Content.Shared.GameObjects.EntitySystems;
+using Content.Shared.Interfaces;
+using Robust.Server.GameObjects;
+using Robust.Server.GameObjects.Components.Container;
+using Robust.Shared.GameObjects;
+using Robust.Shared.Interfaces.GameObjects;
+using Robust.Shared.IoC;
+using Robust.Shared.Localization;
+
+namespace Content.Server.GameObjects.Storage
+{
+    [RegisterComponent]
+    [ComponentReference(typeof(IActivate))]
+    [ComponentReference(typeof(IInteractUsing))]
+    public class WeaponStorageComponent : Component, IActivate, IInteractUsing
+    {
+        protected IEntity _heldItem;
+        protected ContainerSlot _container;
+        protected AppearanceComponent _appearanceComponent;
+        public WallStorageStatus Status => _status;
+        protected WallStorageStatus _status;
+
+        public override string Name => "WeaponStorage";
+
+        bool IInteractUsing.InteractUsing(InteractUsingEventArgs eventArgs)
+        {
+            var result = TryInsertItem(eventArgs.Using);
+            if (!result)
+            {
+                var localizationManager = IoCManager.Resolve<ILocalizationManager>();
+                eventArgs.User.PopupMessage(Owner, localizationManager.GetString("Unable to insert item"));
+            }
+
+            return result;
+        }
+
+        public void RemoveItem(IEntity user)
+        {
+            var heldItem = _container.ContainedEntity;
+            if (heldItem == null)
+            {
+                return;
+            }
+
+            _container.Remove(heldItem);
+            if (user.TryGetComponent(out HandsComponent handsComponent))
+            {
+                handsComponent.PutInHandOrDrop(heldItem.GetComponent<ItemComponent>());
+            }
+
+            if (heldItem.TryGetComponent(out ServerBatteryBarrelComponent batteryBarrelComponent))
+            {
+                batteryBarrelComponent.UpdateAppearance();
+            }
+
+            UpdateStatus();
+        }
+
+        protected abstract WallStorageStatus GetStatus();
```

Comment:
What's the point of this? the status is already publically gettable via `Status => _status` on line 25.

Also if you want derived classes to override this (guessing from the `abstract`), then you want `virtual` not `abstract`